joint_state_controller:
  type: joint_state_controller/JointStateController
  publish_rate: 200
  # 不指定 joints 参数，让控制器自动从硬件接口读取所有注册的关节
  # 这样可以避免关节名称不匹配的问题

# 为 MoveIt 提供 follow_joint_trajectory action 的控制器
# 注意：需要根据实际的 MoveIt 配置中的关节名称来调整
# 这里使用通用的关节名称，实际使用时需要映射到 MoveIt 的关节名称

# 左臂控制器
left_arm_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - LeftShoulderPitch
    - LeftShoulderRoll
    - LeftShoulderYaw
    - LeftElbowPitch
    - LeftElbowYaw
    - LeftWristPitch
    - LeftWristRoll
  gains:
    LeftShoulderPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftShoulderRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftShoulderYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftElbowPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftElbowYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftWristPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftWristRoll: {p: 100, d: 1, i: 1, i_clamp: 1}

# 右臂控制器
right_arm_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - RightShoulderPitch
    - RightShoulderRoll
    - RightShoulderYaw
    - RightElbowPitch
    - RightElbowYaw
    - RightWristPitch
    - RightWristRoll
  gains:
    RightShoulderPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightShoulderRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightShoulderYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightElbowPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightElbowYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightWristPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightWristRoll: {p: 100, d: 1, i: 1, i_clamp: 1}

# 左腿控制器
left_leg_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - LeftHipPitch
    - LeftHipRoll
    - LeftHipYaw
    - LeftKnee
    - LeftAnklePitch
    - LeftAnkleRoll
  gains:
    LeftHipPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftHipRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftHipYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftKnee: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftAnklePitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftAnkleRoll: {p: 100, d: 1, i: 1, i_clamp: 1}

# 右腿控制器
right_leg_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - RightHipPitch
    - RightHipRoll
    - RightHipYaw
    - RightKnee
    - RightAnklePitch
    - RightAnkleRoll
  gains:
    RightHipPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightHipRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightHipYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightKnee: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightAnklePitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightAnkleRoll: {p: 100, d: 1, i: 1, i_clamp: 1}

# 颈部控制器
neck_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - NeckYaw
    - NeckRoll
    - NeckPitch
  gains:
    NeckYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    NeckRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    NeckPitch: {p: 100, d: 1, i: 1, i_clamp: 1}

# 上半身控制器（包含双臂和颈部）
upper_body_controller:
  type: effort_controllers/JointTrajectoryController
  joints:
    - LeftShoulderPitch
    - LeftShoulderRoll
    - LeftShoulderYaw
    - LeftElbowPitch
    - LeftElbowYaw
    - LeftWristPitch
    - LeftWristRoll
    - NeckYaw
    - NeckRoll
    - NeckPitch
    - RightShoulderPitch
    - RightShoulderRoll
    - RightShoulderYaw
    - RightElbowPitch
    - RightElbowYaw
    - RightWristPitch
    - RightWristRoll
  gains:
    LeftShoulderPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftShoulderRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftShoulderYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftElbowPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftElbowYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftWristPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    LeftWristRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    NeckYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    NeckRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    NeckPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightShoulderPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightShoulderRoll: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightShoulderYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightElbowPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightElbowYaw: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightWristPitch: {p: 100, d: 1, i: 1, i_clamp: 1}
    RightWristRoll: {p: 100, d: 1, i: 1, i_clamp: 1}

